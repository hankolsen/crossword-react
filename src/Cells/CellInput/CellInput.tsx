import React, { useEffect, useRef } from 'react';

type Props = {
  clickHandler: () => void;
  inputHandler: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  keyUpHandler: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  left: number;
  width: number;
  height: number;
  top: number;
};

const CellInput = ({
  clickHandler,
  inputHandler,
  keyUpHandler,
  left,
  width,
  height,
  top,
}: Props) => {
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    inputRef.current?.focus();
  });

  return (
    <div
      className="crossword__hidden-input-wrapper"
      style={{
        width: `${width}%`,
        height: `${height}%`,
        top: `${top}%`,
        left: `${left}%`,
      }}
    >
      <input
        type="text"
        aria-label="letter"
        maxLength={1}
        value=""
        autoComplete="off"
        spellCheck="false"
        autoCorrect="off"
        className="crossword__hidden-input"
        onKeyUp={keyUpHandler}
        onInput={inputHandler}
        onClick={clickHandler}
        ref={inputRef}
      />
    </div>
  );
};

export default CellInput;
