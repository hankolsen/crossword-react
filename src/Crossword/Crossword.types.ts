/* istanbul ignore file */
import React from 'react';
import { AxiosError } from 'axios';

export type CellType = {
  [direction in Direction]: string[];
} & {
  column: number;
  number?: number;
  row: number;
  highlighted?: boolean;
  selected?: boolean;
  text?: string;
  arrow?: string;
};

export enum Direction {
  across = 'across',
  down = 'down',
}

export type Selection = {};

export type Position = { x: number; y: number };

export type SeparatorType = {
  direction: Direction;
  position: Position;
  separator: string;
  locations: number[];
  id: string;
};

export type CrosswordListItem = {
  name: string;
  id: string;
};

export type CrosswordListResponse = {
  crosswords: CrosswordListItem[];
};

type Entry = {
  id: string;
  direction: Direction;
  group: string[];
  length: number;
  position: { x: number; y: number };
  separatorLocations?: { [key: string]: number[] };
  turns?: number[];
};

export type CrosswordResponse = {
  crossword: {
    entries: Entry[];
    name: string;
    size: { width: number; height: number };
    _id: string;
  };
};

export type CrosswordType = {
  name: string;
  cells: CellType[][];
  separators?: SeparatorType[];
  numberOfColumns: number;
  numberOfRows: number;
  boardWidth: number;
  boardHeight: number;
  inputWidth: number;
  inputHeight: number;
  crosswordId: string;
};

export type CrosswordContextType = {
  boardWidth: number;
  boardHeight: number;
  clickHandler: (row: number, column: number) => void;
  cells?: CellType[][];
  crosswordId: string;
  currentCell: CellType;
  inputWidth: number;
  inputHeight: number;
  isLoading: boolean;
  name?: string;
  numberOfColumns?: number;
  numberOfRows?: number;
  inputClickHandler: () => void;
  inputHandler: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  keyUpHandler: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  reset: () => void;
  separators?: SeparatorType[];
  showInput: boolean;
  error?: AxiosError;
};
