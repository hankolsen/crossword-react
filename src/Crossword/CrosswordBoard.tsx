import React from 'react';
import CellInput from 'Cells/CellInput/CellInput';
import Cells from 'Cells/Cells';
import Separators from 'Separators/Separators';
import { useCrossword } from './CrosswordContext';

const CrosswordBoard = () => {
  const { boardWidth, boardHeight, separators } = useCrossword();
  return (
    <div className="crossword-board">
      <svg
        className="crossword__grid"
        viewBox={`0 0 ${boardWidth} ${boardHeight}`}
        fill="#222222"
      >
        <rect
          x="0"
          y="0"
          width={boardWidth}
          height={boardHeight}
          className="crossword__grid-background"
        />
        <Cells />
        <Separators separators={separators} />
      </svg>
      <CellInput />
    </div>
  );
};

export default CrosswordBoard;
