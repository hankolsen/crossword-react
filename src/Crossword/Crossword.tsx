import React from 'react';
import LoadingIndicator from '../LoadingIndicator/LoadingIndicator';
import CrosswordBoard from './CrosswordBoard';
import { useCrossword } from './CrosswordContext';

import './Crossword.css';
import ResetButton from './ResetButton';

const Crossword = () => {
  const { isLoading, name } = useCrossword();

  if (isLoading) {
    return (
      <div className="crossword">
        <h3>Loading crossword</h3>
        <LoadingIndicator />
      </div>
    );
  }

  return (
    <div className="crossword">
      <h3>{name}</h3>
      <div className="crossword-container">
        <CrosswordBoard />
      </div>
      <ResetButton />
    </div>
  );
};

export default Crossword;
