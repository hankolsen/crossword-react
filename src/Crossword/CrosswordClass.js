/* eslint-disable no-param-reassign */
import React from 'react';
import PropTypes from 'prop-types';

import './Crossword.css';
import { Redirect } from 'react-router-dom';
import Separators from '../Separators/Separators';
import Cells from '../Cells/Cells';
import CellInput from '../Cells/CellInput/CellInput';
import LoadingIndicator from '../LoadingIndicator/LoadingIndicator';
import {
  cellContainsOtherDirection,
  createCrossword,
  cellIsStartingWord,
  deselectAll,
  emptyAll,
  getCurrentId,
  getInputPosition,
  highlightId,
  isIgnorableKey,
  isValidKey,
  toggleDirection,
} from './crosswordHelper';

class Crossword extends React.Component {
  constructor(props) {
    super(props);

    this.separators = [];
    this.numberOfColumns = 0;
    this.numberOfRows = 0;
    this.inputWidth = 0;
    this.inputHeight = 0;
    this.boardWidth = 0;
    this.boardHeight = 0;

    this.state = {
      isLoading: true,
      direction: 'across',
      error: false,
      showInput: false,
    };

    this.clickHandler = this.clickHandler.bind(this);
    this.inputHandler = this.inputHandler.bind(this);
    this.keyUpHandler = this.keyUpHandler.bind(this);
    this.inputClickHandler = this.inputClickHandler.bind(this);
    this.reset = this.reset.bind(this);
  }

  componentDidMount() {
    const { match } = this.props;
    this.crosswordId = match.params.id;
    createCrossword(this.crosswordId)
      .then(
        ({
          name,
          cells,
          separators,
          numberOfColumns,
          numberOfRows,
          boardWidth,
          boardHeight,
          inputWidth,
          inputHeight,
        }) => {
          Object.assign(this, {
            name,
            separators,
            numberOfColumns,
            numberOfRows,
            boardWidth,
            boardHeight,
            inputWidth,
            inputHeight,
          });

          this.setState({ cells });
        },
      )
      .catch(() => {
        this.setState({ error: true });
      })
      .finally(() => this.setState({ isLoading: false }));
  }

  handleArrowMove({ arrow }) {
    const { cells, currentCell } = this.state;
    const { row, column } = currentCell;

    /* eslint-disable no-shadow */
    const moves = {
      Left: (cells, row, column) => cells[row][column - 1],
      Right: (cells, row, column) => cells[row][column + 1],
      Up: (cells, row, column) => cells[row - 1] && cells[row - 1][column],
      Down: (cells, row, column) => cells[row + 1] && cells[row + 1][column],
    };
    /* eslint-enable no-shadow */

    const newCell = moves[arrow](cells, row, column);
    if (newCell) {
      this.moveTo(newCell);
    }
  }

  inputClickHandler() {
    let { direction, currentCell } = this.state;
    const { cells, selection } = this.state;

    if (currentCell) {
      if (cellContainsOtherDirection({ currentCell, direction })) {
        direction = toggleDirection(direction);
        this.highlightCurrentSelection({ direction });
      } else if (currentCell[direction].length > 1) {
        const currentIndex = currentCell[direction].indexOf(selection);
        const nextIndex = (currentIndex + 1) % currentCell[direction].length;
        const newId = currentCell[direction][nextIndex];
        highlightId({ cells, direction, id: newId, currentCell });
        this.setState({ cells, selection: newId });
      }
    } else {
      currentCell = cells[0][0]; // eslint-disable-line prefer-destructuring
      if (!currentCell[direction]) {
        direction = toggleDirection(direction);
      }
      this.setState({ currentCell, direction }, () => {
        this.highlightCurrentSelection({ direction });
      });
    }
  }

  clickHandler(event, row, column) {
    let { direction } = this.state;
    const { cells } = this.state;
    const currentCell = cells[row][column];
    if (
      currentCell.number &&
      !currentCell.highlighted &&
      !cellIsStartingWord({ cell: currentCell, direction }) &&
      cellContainsOtherDirection({ currentCell, direction }) &&
      cellIsStartingWord({
        cell: currentCell,
        direction: toggleDirection(direction),
      })
    ) {
      direction = toggleDirection(direction);
    }
    if (!currentCell[direction] || currentCell.selected) {
      direction = toggleDirection(direction);
    }
    this.setState({ showInput: true });
    this.setState({ currentCell }, () => {
      this.highlightCurrentSelection({ direction }).then(
        ({ direction: newDirection }) => {
          this.setState({ cells, direction: newDirection, currentCell });
        },
      );
    });
  }

  highlightCurrentSelection({ direction }) {
    const { cells, currentCell, selection } = this.state;
    return new Promise((resolve) => {
      const id = getCurrentId({ currentCell, direction, selection });
      if (id) {
        highlightId({ cells, direction, id, currentCell });
      }
      this.setState({ direction, selection: id }, () => resolve({ direction }));
    });
  }

  reset() {
    localStorage.removeItem(`kryzz-${this.crosswordId}`);
    const { cells } = this.state;
    emptyAll(cells);
    this.setState({ cells });
  }

  inputHandler(event) {
    const { currentCell } = this.state;
    const key = event.target.value;
    if (isValidKey(key)) {
      currentCell.text = key.toUpperCase();
      this.setState({ currentCell });
      this.moveToNext();
    }
  }

  keyUpHandler(event) {
    const { key } = event;

    if (key === 'Shift' || event.metaKey) {
      return;
    }

    // Is it an arrow key?
    const [, arrow] = key.match(/Arrow(\w+)$/) || [];

    if (arrow) {
      this.handleArrowMove({ arrow });
    }

    if (isIgnorableKey(key)) {
      event.preventDefault();
      return;
    }

    if (key === 'Backspace') {
      const { currentCell } = this.state;
      if (currentCell.text) {
        currentCell.text = '';
        this.setState({ currentCell });
      } else {
        this.moveToPrevious();
      }
    }

    const { cells } = this.state;
    const entries = cells.map((row) => row.map((cell) => cell && cell.text));
    localStorage.setItem(`kryzz-${this.crosswordId}`, JSON.stringify(entries));
  }

  moveTo(nextCell) {
    const { cells } = this.state;
    let { direction, selection } = this.state;
    if (!nextCell[direction] || !nextCell[direction].includes(selection)) {
      direction = toggleDirection(direction);
      if (!nextCell[direction] || !nextCell[direction].length) {
        direction = toggleDirection(direction);
      }
      [selection] = nextCell[direction];
    }
    highlightId({ cells, direction, id: selection, currentCell: nextCell });
    this.setState({ cells, selection, currentCell: nextCell, direction });
  }

  moveToPrevious() {
    this.moveToNext(-1);
  }

  moveToNext(dir = 1) {
    const { direction, cells, selection } = this.state;
    const array = [];

    cells.map((row) =>
      row.forEach((cell) => {
        if (cell && cell[direction] && cell[direction].includes(selection)) {
          array.push(cell);
        }
      }),
    );

    let nextCell;
    const currentIndex = array.findIndex((cell) => cell.selected);
    if (currentIndex > -1 && currentIndex < array.length) {
      const nextIndex = currentIndex + dir;
      nextCell = array[nextIndex];
    }

    if (nextCell) {
      deselectAll(cells);
      nextCell.selected = true;
      this.setState({ currentCell: nextCell, cells });
    }
  }

  renderCrossword() {
    const { cells, currentCell = {}, showInput } = this.state;
    const { left, top } = getInputPosition(currentCell);
    return (
      <div className="crossword-board">
        <svg
          className="crossword__grid"
          viewBox={`0 0 ${this.boardWidth} ${this.boardHeight}`}
          fill="#222222"
        >
          <rect
            x="0"
            y="0"
            width={this.boardWidth}
            height={this.boardHeight}
            className="crossword__grid-background"
          />
          {cells && <Cells cells={cells} clickHandler={this.clickHandler} />}
          <Separators separators={this.separators} />
        </svg>
        {showInput && (
          <CellInput
            top={top}
            left={left}
            width={this.inputWidth}
            height={this.inputHeight}
            inputHandler={this.inputHandler}
            keyUpHandler={this.keyUpHandler}
            clickHandler={this.inputClickHandler}
          />
        )}
      </div>
    );
  }

  render() {
    const { isLoading, error } = this.state;

    if (error) {
      return <Redirect to="/" />;
    }

    return (
      <div className="crossword">
        <h3>{this.name ? this.name : 'Loading crossword'}</h3>
        <div className="crossword-container">
          {isLoading ? <LoadingIndicator /> : this.renderCrossword()}
        </div>
        <button
          onClick={this.reset}
          type="submit"
          className={isLoading ? 'hidden' : ''}
        >
          Reset
        </button>
      </div>
    );
  }
}

Crossword.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string,
    }).isRequired,
  }).isRequired,
};

export default Crossword;
