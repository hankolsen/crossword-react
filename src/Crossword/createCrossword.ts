import { CELL_HEIGHT, CELL_WIDTH } from '../config';
import { CellType } from './Crossword.types';
import { toggleDirection } from './crosswordHelper';
import fillCell from './fillCell';

type Entry = {
  id: string;
  direction: 'across' | 'down';
  group: string[];
  length: number;
  position: { x: number; y: number };
  separatorLocations?: { [key: string]: number[] };
  turns?: number[];
};

export type CrosswordType = {
  name: string;
  cells: CellType[][];
  separators?: unknown[];
  numberOfColumns: number;
  numberOfRows: number;
  boardWidth: number;
  boardHeight: number;
  inputWidth: number;
  inputHeight: number;
};

type CrosswordResponse = {
  crossword: {
    entries: Entry[];
    name: string;
    size: { width: number; height: number };
    _id: string;
  };
};

const createCrossword = (crosswordId: string): Promise<CrosswordType> =>
  new Promise((resolve, reject) =>
    fetch(`/.netlify/functions/crosswords/${crosswordId}`)
      .then((response) => response.json())
      .then(({ crossword }: CrosswordResponse) => {
        if (!crossword) {
          reject();
        }
        const numberOfColumns = crossword.size.width;
        const numberOfRows = crossword.size.height;
        const separators: unknown[] = [];

        const cells = Array(numberOfRows)
          .fill(null)
          .map(() => Array(numberOfColumns).fill(null));

        const userData = JSON.parse(
          localStorage.getItem(`kryzz-${crosswordId}`) || 'null',
        );

        crossword.entries.sort(
          (a, b) => a.position.y - b.position.y || a.position.x - b.position.x,
        );

        let number = 1;
        const numbers: { [key: string]: number } = {};
        crossword.entries.forEach(
          ({ direction, position, length, separatorLocations, turns }) => {
            let column = position.x;
            let row = position.y;
            let walkingDirection = direction;
            let turnIndex = 0;
            if (numbers[`${column},${row}`]) {
              number = numbers[`${column},${row}`];
            }
            const id = `${number}-${direction}`;
            numbers[`${column},${row}`] = number;
            Array(length)
              .fill(null)
              .forEach((_, i) => {
                const text = userData && userData[row][column];
                fillCell({
                  cells,
                  row,
                  column,
                  index: i,
                  number,
                  id,
                  direction,
                  text,
                  hasTurn: turns,
                });

                if (turns && turns.length && turns[turnIndex] - 1 === i) {
                  walkingDirection = toggleDirection(walkingDirection);
                  cells[row][column].arrow = walkingDirection;
                  turnIndex += 1;
                }

                if (walkingDirection === 'across') {
                  column += 1;
                } else {
                  row += 1;
                }
              });

            Object.entries(separatorLocations || {}).forEach(
              ([separator, locations]) => {
                if (locations && locations.length) {
                  separators.push({
                    direction,
                    position,
                    separator,
                    locations,
                    id,
                  });
                }
              },
            );
            number += 1;
          },
        );

        const inputWidth = 100 / numberOfColumns;
        const inputHeight = 100 / numberOfRows;
        const boardWidth =
          CELL_WIDTH * numberOfColumns + numberOfColumns + 1 || 0;
        const boardHeight = CELL_HEIGHT * numberOfRows + numberOfRows + 1 || 0;
        const { name } = crossword;
        console.log('cells', cells);
        // eslint-disable-next-line max-len
        resolve({
          name,
          cells,
          separators,
          numberOfColumns,
          numberOfRows,
          boardWidth,
          boardHeight,
          inputWidth,
          inputHeight,
        });
      })
      .catch((e) => reject(e)),
  );

export default createCrossword;
