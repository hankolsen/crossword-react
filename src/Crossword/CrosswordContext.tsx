/* istanbul ignore file */
import React, {
  createContext,
  FunctionComponent,
  useContext,
  useEffect,
  useReducer,
} from 'react';
import { useParams } from 'react-router-dom';
import { AxiosError } from 'axios';
import { CellType, CrosswordContextType, Direction } from './Crossword.types';
import crosswordReducer from './crosswordReducer';
import createCrossword from './utils/createCrossword';
import useSingleCrossword from './useSingleCrossword';

const CrosswordContext = createContext<CrosswordContextType | undefined>(
  undefined,
);

const useCrossword = () => {
  const context = useContext(CrosswordContext);
  if (!context) {
    throw new Error('useCrossword must be used within a CrosswordContext');
  }
  return context;
};

const CrosswordProvider: FunctionComponent = ({ children }) => {
  const initialState = {
    crossword: undefined,
    showInput: false,
    selection: '',
    currentCell: { row: 0, column: 0 } as CellType,
    direction: Direction.across,
    error: undefined,
  };
  const [{ crossword, currentCell, showInput }, dispatch] = useReducer(
    crosswordReducer,
    initialState,
  );
  const { crosswordId } = useParams<{ crosswordId: string }>();

  const clickHandler = (row: number, column: number) => {
    dispatch({ type: 'click_cell', row, column });
  };

  const inputHandler = (event: React.KeyboardEvent<HTMLInputElement>) => {
    dispatch({ type: 'on_input', event });
  };

  const keyUpHandler = (event: React.KeyboardEvent<HTMLInputElement>) =>
    dispatch({ type: 'key_up', event });

  const inputClickHandler = () => dispatch({ type: 'click_input' });

  const reset = () => dispatch({ type: 'reset' });

  const { data, isLoading, error } = useSingleCrossword(crosswordId);

  useEffect(() => {
    if (data) {
      const crossword = createCrossword({
        crossword: data.crossword,
        crosswordId,
      });
      dispatch({ type: 'crossword_loaded', crossword });
    }
  }, [crosswordId, data]);

  const {
    cells,
    numberOfColumns,
    numberOfRows,
    boardWidth = 0,
    boardHeight = 0,
    inputWidth = 0,
    inputHeight = 0,
    name,
    separators,
  } = crossword ?? {};

  const value: CrosswordContextType = {
    boardWidth,
    boardHeight,
    clickHandler,
    cells,
    currentCell,
    crosswordId,
    inputWidth,
    inputHeight,
    isLoading,
    name,
    numberOfColumns,
    numberOfRows,
    inputClickHandler,
    inputHandler,
    keyUpHandler,
    separators,
    showInput,
    reset,
    error: error as AxiosError,
  };

  return (
    <CrosswordContext.Provider value={value}>
      {children}
    </CrosswordContext.Provider>
  );
};

export { CrosswordProvider, useCrossword, CrosswordContext };
