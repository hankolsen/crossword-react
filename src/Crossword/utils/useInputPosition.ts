import { useCrossword } from 'Crossword/CrosswordContext';

const useInputPosition = () => {
  const { currentCell, numberOfRows = 1, numberOfColumns = 1 } = useCrossword();
  const { row = 0, column = 0 } = currentCell;
  const top = (row / numberOfRows) * 100 || 0; // (((row * CELL_HEIGHT) + 2) / boardHeight) * 100 || 0;
  const left = (column / numberOfColumns) * 100 || 0;

  return { left, top };
};

export default useInputPosition;
