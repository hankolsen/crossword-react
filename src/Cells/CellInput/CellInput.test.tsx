import React from 'react';
import { render, screen } from 'test/test-utils';
import { useCrossword } from 'Crossword/CrosswordContext';
import useInputPosition from 'Crossword/utils/useInputPosition';
import CellInput from './CellInput';

jest.mock('Crossword/CrosswordContext');
const mockUseCrossword = useCrossword as jest.Mock;

jest.mock('Crossword/utils/useInputPosition');
const mockUseInputPosition = useInputPosition as jest.Mock;

describe('CellInput test', () => {
  it('should not render if not show is true', () => {
    mockUseCrossword.mockImplementation(() => ({
      showInput: false,
    }));
    mockUseInputPosition.mockImplementation(() => ({}));

    const { container } = render(<CellInput />);
    expect(
      container.querySelector('.crossword__hidden-input-wrapper'),
    ).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render an input', () => {
    mockUseCrossword.mockImplementation(() => ({
      showInput: true,
      inputHandler: () => {},
    }));

    mockUseInputPosition.mockImplementation(() => ({}));

    const { container } = render(<CellInput />);
    expect(
      container.querySelector('.crossword__hidden-input-wrapper'),
    ).toBeInTheDocument();
    expect(screen.queryByLabelText('letter')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
    expect(container.querySelector('.crossword__hidden-input-wrapper'));
  });
});
