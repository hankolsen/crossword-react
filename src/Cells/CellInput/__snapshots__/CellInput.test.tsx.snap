// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellInput test should not render if not show is true 1`] = `
<div>
  <div />
</div>
`;

exports[`CellInput test should render an input 1`] = `
<div>
  <div>
    <div
      class="crossword__hidden-input-wrapper"
    >
      <input
        aria-label="letter"
        autocomplete="off"
        autocorrect="off"
        class="crossword__hidden-input"
        maxlength="1"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
